sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/query/GroupingTable.java

```diff
@@ -0,0 +1,69 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+package com.azure.cosmos.implementation.query;
+
+import com.azure.cosmos.implementation.Document;
+import com.azure.cosmos.implementation.query.aggregation.AggregateOperator;
+import com.azure.cosmos.implementation.routing.UInt128;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class GroupingTable {
+    private static final List<AggregateOperator> EmptyAggregateOperators = new ArrayList<>();
+
+    private final Map<UInt128, SingleGroupAggregator> table;
+    private final Map<String, AggregateOperator> groupByAliasToAggregateType;
+    private final List<String> orderedAliases;
+    private final boolean hasSelectValue;
+
+    GroupingTable(Map<String, AggregateOperator> groupByAliasToAggregateType, List<String> orderedAliases,
+        boolean hasSelectValue) {
+        this.table = new HashMap<>();
+        this.groupByAliasToAggregateType = groupByAliasToAggregateType;
+        this.orderedAliases = orderedAliases;
+        this.hasSelectValue = hasSelectValue;
+    }
+
+    public void addPayLoad(GroupByDocumentQueryExecutionContext<?>.RewrittenGroupByProjection rewrittenGroupByProjection) {
+        try {
+            final UInt128 groupByKeysHash = DistinctHash.getHash(rewrittenGroupByProjection.getGroupByItems());
+            SingleGroupAggregator singleGroupAggregator;
+            if (!this.table.containsKey(groupByKeysHash)) {
+                singleGroupAggregator = SingleGroupAggregator.create(EmptyAggregateOperators,
+                                                                     this.groupByAliasToAggregateType,
+                                                                     this.orderedAliases,
+                                                                     this.hasSelectValue,
+                                                                      /*continuationtoken*/ null);
+                this.table.put(groupByKeysHash, singleGroupAggregator);
+            } else {
+                singleGroupAggregator = table.get(groupByKeysHash);
+            }
+
+            singleGroupAggregator.addValues(rewrittenGroupByProjection.getPayload());
+
+        } catch (IOException e) {
+            throw new IllegalStateException("Failed to add payload to groupby projection", e);
+        }
+    }
+
+    public List<Document> drain(int maxItemCount) {
+        Collection<UInt128> keys = this.table.keySet().stream().limit(maxItemCount).collect(Collectors.toList());
```

Comment:
Since there is no ordering in the `keySet()` API - are we just limiting with `maxItemCount` randomly on the keys in `this.table` map ?

Reply:
The ordering doesn't matter in this scenario. The user is just getting back some subset of the keys. It's never defined what order that would be.